### PR TITLE
Nme gpu

### PR DESCRIPTION
**Pull request by wghezaiel**
Eigenvalues decomposition is computed with "torch.linalg.eigh" on GPU

### CODE REVIEW

File: simple/diarization/processing/simple_diarizer/simple_diarizer/diarizer.py [hunk 1]
```diff
@@ -56,7 +56,7 @@ def __init__(
         if not num_threads:
             num_threads = torch.get_num_threads()
 
-        self.log(f"Devices: VAD={device_vad}, embedding={device}, clustering=cpu (with {num_threads} CPU threads)")
+        self.log(f"Devices: VAD={device_vad}, embedding={device}, clustering={device} (with {num_threads} CPU threads)")
```
Comment (Jeronymous) [inline simple/diarization/processing/simple_diarizer/simple_diarizer/diarizer.py:59]: please introduce another variable `device_clustering` (same as `device_vad`).
Because in practice it might be relevant to run GPU **only** for embeddings (and until we have an emperical experience, better to let every detail configurable)

File: simple/diarization/processing/simple_diarizer/simple_diarizer/spectral_clustering.py [hunk 2]
```diff
@@ -38,33 +39,42 @@ def Eigengap(S):
...
-        mode="buckling",
-    )
-    # Get largest eigenvalue
-    Smax = scipy.sparse.linalg.eigsh(
-        Lp, k=1, which="LA", tol=1e-6, return_eigenvectors=False, mode="buckling"
-    )
+    from torch.linalg import eigh     
+    
+    if device=="cuda" and torch.cuda.is_available()== True:       
+        Lp = torch.from_numpy(Lp).float().to('cuda')
```
Comment (Jeronymous) [inline simple/diarization/processing/simple_diarizer/simple_diarizer/spectral_clustering.py:58]: Be careful with shaky test:
* device=="cuda" is not exhaustive (it can be "cuda:0", ...)
* It is not the place where to test "torch.cuda.is_available()" (if the users specify "cuda" explicitly, he wants to run it on CUDA or fail)

These 2 lines can be more robust (and general) with:
```
if device != "cpu":
   Lp = torch.from_numpy(Lp).float().to(device)
```
(not the second use of "device",important to correctly use torch)

File: simple/diarization/processing/simple_diarizer/simple_diarizer/spectral_clustering.py [hunk 3]
```diff
@@ -38,33 +39,42 @@ def Eigengap(S):
...
     # Laplacian matrix computation
     Lp = Laplacian(Ap)
     # Get max_num_clusters+1 smallest eigenvalues
-    S = scipy.sparse.linalg.eigsh(
-        Lp,
-        k=max_num_clusters + 1,
-        which="SA",
-        tol=1e-6,
-        return_eigenvectors=False,
-        mode="buckling",
```
Comment (Jeronymous) [inline simple/diarization/processing/simple_diarizer/simple_diarizer/spectral_clustering.py:57]: Those settings do not seem to have been kept in the new version.

Have you checked that all intermediate values (e, g, k, r) remain the same on a bunch of example inputs.

File: simple/diarization/processing/simple_diarizer/simple_diarizer/spectral_clustering.py [hunk 4]
```diff
@@ -38,33 +39,42 @@ def Eigengap(S):
...
-    r = p / g
-    k = np.argmax(e[:max_num_clusters])
+    
+    e = np.sort(S)
+    g = getLamdaGaplist(e)
+    k = np.argmax(g[: min(max_num_clusters, len(g))]) 
+    arg_sorted_idx = np.argsort(g[: max_num_clusters])[::-1]
+    max_key = arg_sorted_idx[0]
+    max_eig_gap = g[max_key] / (max(e) + 1e-10)
+    r = (p / A.shape[0]) / (max_eig_gap + 1e-10)
```
Comment (Jeronymous) [inline simple/diarization/processing/simple_diarizer/simple_diarizer/spectral_clustering.py:75]: This diff is hard to follow. There are some sort/argsort added, operations are not done in the same order...

Can you explain, please, so that we check if (e, g, k, r) remain the same quantities.

Maybe you can start by explaining why Smax disappeared?
Were the previous "S" and "Smax" the extreme eigenvalues of what is now the list of eigenvalues S (sorted in e)?

File: simple/diarization/processing/simple_diarizer/simple_diarizer/spectral_clustering.py [hunk 5]
```diff
@@ -81,23 +91,22 @@ def ComputeNMEParameters(A, p, max_num_clusters):
...
-    A, num_clusters=None, max_num_clusters=None, pbest=0, pmin=3, pmax=20
+    A, num_clusters=None, max_num_clusters=None, pbest=0, pmin=3, pmax=20, device=None
 ):
     if max_num_clusters is None:
         assert num_clusters is not None, "Cannot have both num_clusters and max_num_clusters be None"
         max_num_clusters = num_clusters
 
     if pbest == 0:
-        # Selecting best number of neighbors for affinity matrix thresolding
+        print("Selecting best number of neighbors for affinity matrix thresolding:")
```
Comment (Jeronymous) [inline simple/diarization/processing/simple_diarizer/simple_diarizer/spectral_clustering.py:101]: Hmm, I am not a big fan of this print.
Logging should be done outside of this function.

If you really want to keep a debug print, at least remove the ":" at the end (there is nothing printed after) and pass  logger as an argument that can be None (if the logger is not None, use it to print)

**Review by Jeronymous — CHANGES_REQUESTED** [review]

File: simple/diarization/processing/simple_diarizer/simple_diarizer/diarizer.py (hunk 1 above)
Comment (wghezaiel) [reply simple/diarization/processing/simple_diarizer/simple_diarizer/diarizer.py:59]: corrected

File: simple/diarization/processing/simple_diarizer/simple_diarizer/spectral_clustering.py (hunk 2 above)
Comment (wghezaiel) [reply simple/diarization/processing/simple_diarizer/simple_diarizer/spectral_clustering.py:58]: thanks, done

File: simple/diarization/processing/simple_diarizer/simple_diarizer/spectral_clustering.py (hunk 3 above)
Comment (wghezaiel) [reply simple/diarization/processing/simple_diarizer/simple_diarizer/spectral_clustering.py:57]: config in torch.linalg.eigh is different to scipy.
verified and same performance are achived

File: simple/diarization/processing/simple_diarizer/simple_diarizer/spectral_clustering.py (hunk 4 above)
Comment (wghezaiel) [reply simple/diarization/processing/simple_diarizer/simple_diarizer/spectral_clustering.py:75]: same performance to last code with scipy. 
From Nemo:
https://github.com/NVIDIA/NeMo/blob/c7a5a33b37c2202e1e0af7ad0eafa7b9c68f1367/nemo/collections/asr/parts/utils/nmesc_clustering.py

File: simple/diarization/processing/simple_diarizer/simple_diarizer/spectral_clustering.py (hunk 5 above)
Comment (wghezaiel) [reply simple/diarization/processing/simple_diarizer/simple_diarizer/spectral_clustering.py:101]: corrected

File: simple/diarization/processing/simple_diarizer/simple_diarizer/spectral_clustering.py (hunk 4 above)
Comment (wghezaiel) [reply simple/diarization/processing/simple_diarizer/simple_diarizer/spectral_clustering.py:75]: we want from (e, g, k, r) only k+1 to predict number of speaker, pbest=p to compute NME_SpectralClustering_sklearn( A, num_clusters, pbest).
p and k are simular to last code with scipy, but e and g are different
